src/backend/access/appendonly/appendonlywriter.c

```diff
@@ -778,6 +778,54 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
 	return;
 }
 
+void
+UpdateSegnoAfterCompaction(Oid relid, List *compactedSegmentFileList)
+{
+	AORelHashEntryData *aoentry;
+	Relation	aosegrel;
+	bool	   *awaiting_drop = NULL;
+	int			i;
+
+	Assert(Gp_role != GP_ROLE_EXECUTE);
+	if (Gp_role == GP_ROLE_UTILITY)
+	{
+		return;
+	}
+
+	if (compactedSegmentFileList == NIL)
+	{
+		return;
+	}
+
+	aosegrel = relation_open(relid, AccessShareLock);
+	awaiting_drop = get_awaiting_drop_status_from_segments(aosegrel);
+	relation_close(aosegrel, AccessShareLock);
+
+	acquire_lightweight_lock();
+
+	aoentry = AORelGetOrCreateHashEntry(relid);
+	Assert(aoentry);
+
+	for (i = 0; i < MAX_AOREL_CONCURRENCY; i++)
+	{
+		AOSegfileStatus *segfilestat = &aoentry->relsegfiles[i];
+
+		if (list_member_int(compactedSegmentFileList, i) && !awaiting_drop[i])
+		{
+			ereportif(Debug_appendonly_print_segfile_choice, LOG,
+					  (errmsg("Update segno %d after compaction "
+							  "relation \"%s\" (%d)", i,
+							  get_rel_name(relid), relid)));
+
+			segfilestat->state = DROP_USE;
+		}
+	}
+
+	release_lightweight_lock();
+	pfree(awaiting_drop);
+	return;
```

Comment:
return at the end of function is redundant

Reply:
Yes, but see line https://github.com/arenadata/gpdb/pull/548/files#diff-9fecdbbedbe0a8ed6898b84ef38c28e23ae4a8f4d6f09b38f57be565c4ee8504R778

Reply:
I don't know what the author means. May be he thinks that a function with return statement at the end and the same function without it work differently. But I don't know the difference. So I suggest to remove redundant return

Reply:
fixed